packages/core/src/mpi/match-patients.ts

```diff
@@ -0,0 +1,167 @@
+  threshold: number
+) => boolean;
+
+type MatchingRule = (patient1: PatientData, patient2: PatientData) => boolean;
+
+/**
+ * `matchPatients` filters patients based on a similarity function, rules, and a threshold.
+ * @param similarityFunction - Determines if a patient is a match.
+ * @param matchingRules - Rules that determines if a patient is a match.
+ * @param patients - Array of patients.
+ * @param demo - Patient data to find matches for.
+ * @param threshold - Minimum similarity score for a match.
+ * @param greedy - If true, returns the first match. If false, returns all matches.
+ * @returns matched patients.
+ */
+export function matchPatients(
+  isSimilarEnough: SimilarityFunction,
+  matchingRules: MatchingRule[],
+  patients: PatientMPI[],
+  currentPatient: PatientData,
+  threshold: number,
+  greedy = true
+): PatientMPI[] {
+  const matchFunction = (patient: PatientMPI) => {
+    const normalizedPatient = normalizePatient(patient);
+    if (!normalizedPatient) {
+      return false;
+    }
+    for (const matchingRule of matchingRules) {
+      if (matchingRule(currentPatient, normalizedPatient)) {
+        return true;
+      }
+    }
+    return isSimilarEnough(normalizedPatient, currentPatient, threshold);
+  };
+  if (greedy) {
+    const foundPatient = patients.find(matchFunction);
+    return foundPatient ? [foundPatient] : [];
+  } else {
+    return patients.filter(matchFunction);
+  }
+}
+
+/**
+ * This function checks if the patient has any personal identifiers that match the demo. The idea of having rules
+ * is that rules will override statistical similarity methods. This rule should act as an override.
+ * @param demo
+ * @param patient
+ * @returns true if the patient has any personal identifiers that match the demo.
+ */
+export function matchingPersonalIdentifiersRule(
+  patient1: PatientData,
+  patient2: PatientData
+): boolean {
+  const identifiers1 = patient1.personalIdentifiers || [];
+  const identifiers2 = patient2.personalIdentifiers || [];
+  const isMatchIdentifier =
+    intersectionWith(identifiers1, identifiers2, isSameIdentifierById).length > 0;
+  return isMatchIdentifier;
+}
+
+/**
+ * This function checks if the patient has any contact details (phone or email) that match the demo. The idea of this rule
+ * us that contact info is a very strong unique identifier if they do match, but doesn't necessarily mean that the patients
+ * don't match if they don't match.
+ * @param demo
+ * @param patient
+ * @returns true if the patient has any contact details that match the demo.
+ */
+export function matchingContactDetailsRule(patient1: PatientData, patient2: PatientData): boolean {
+  const contact1 = patient1.contact || [];
+  const contact2 = patient2.contact || [];
+  const isMatchPhone = intersectionWith(contact1, contact2, isSameContactByPhone).length > 0;
+  const isMatchEmail = intersectionWith(contact1, contact2, isSameContactByEmail).length > 0;
+  return isMatchPhone || isMatchEmail;
```

Comment:
nit: since these can potentially be executed millions of times, it's worth considering small improvements to save some CPU cycles - especially when its a simple change.

In this case, we could return `true` if `isMatchPhone` is true, otherwise we check the email as well.

Reply:
Haha thats what I had initially! https://github.com/metriport/metriport/pull/1160#discussion_r1437865253
But it was tough to read.

Reply:
It's a stretch to say that code is the only solution to what I said here, right?

Here you go:

```ts
  const contact1 = patient1.contact || [];
  const contact2 = patient2.contact || [];
  const isMatchPhone = intersectionWith(contact1, contact2, isSameContactByPhone).length > 0;
  if (isMatchPhone) return true;
  const isMatchEmail = intersectionWith(contact1, contact2, isSameContactByEmail).length > 0;
  return isMatchEmail;
```

Reply:
I implemented the exact solution you suggested there, but I agree that the code you suggested here is probably the best solution that is both readable and efficient